ravelin.js

```diff
@@ -1,1185 +1,1743 @@
-      BigInteger.prototype.am = am3;
-      dbits = 28;
+
+    for (var prop in details) {
+      if (!details.hasOwnProperty(prop)) continue;
+      switch (prop) {
+        case 'pan':
+        case 'year':
+        case 'month':
+        case 'nameOnCard':
+          continue;
+      }
+      throw new Error('[ravelinjs] Encryption validation: encrypt only allows properties pan, year, month, nameOnCard');
     }
 
-    BigInteger.prototype.DB = dbits;
-    BigInteger.prototype.DM = ((1<<dbits)-1);
-    BigInteger.prototype.DV = (1<<dbits);
+    details.month += '';
+    details.year += '';
 
-    var BI_FP = 52;
-    BigInteger.prototype.FV = Math.pow(2,BI_FP);
-    BigInteger.prototype.F1 = BI_FP-dbits;
-    BigInteger.prototype.F2 = 2*dbits-BI_FP;
+    // AES encrypt the card details, using a uniquely generated session key and IV.
+    // The GCM auth tag is appended to the ciphertext. This key and IV are returned from this call as well
+    // as the card detail ciphertext, all as base64.
+    var aesResult = aesEncrypt(JSON.stringify(details));
 
-    // Digit conversions
-    var BI_RM = "0123456789abcdefghijklmnopqrstuvwxyz";
-    var BI_RC = new Array();
-    var rr,vv;
-    rr = "0".charCodeAt(0);
-    for(vv = 0; vv <= 9; ++vv) BI_RC[rr++] = vv;
-    rr = "a".charCodeAt(0);
-    for(vv = 10; vv < 36; ++vv) BI_RC[rr++] = vv;
-    rr = "A".charCodeAt(0);
-    for(vv = 10; vv < 36; ++vv) BI_RC[rr++] = vv;
+    // RSA encrypt the key and IV from the previous step, as a single, pipe-delimited string.
+    var rsaResultB64 = rsaEncrypt(this.rsaKey, aesResult.aesKeyB64, aesResult.ivB64);
 
-    function int2char(n) { return BI_RM.charAt(n); }
-    function intAt(s,i) {
-      var c = BI_RC[s.charCodeAt(i)];
-      return (c==null)?-1:c;
-    }
+    // This payload identically matches the structure we expect to be sent to the Ravelin API
+    return {
+      methodType: 'paymentMethodCipher',
+      cardCiphertext: aesResult.ciphertextB64,
+      aesKeyCiphertext: rsaResultB64,
+      algorithm: 'RSA_WITH_AES_256_GCM',
+      ravelinSDKVersion: FULL_VERSION_STRING,
+      keyIndex: this.keyIndex
+    };
+  }
 
-    // (protected) copy this to r
-    function bnpCopyTo(r) {
-      for(var i = this.t-1; i >= 0; --i) r[i] = this[i];
-      r.t = this.t;
-      r.s = this.s;
-    }
+  /**
+   * trackFingerprint sends device information back to Ravelin. Invoke from
+   * the checkout page of your payment flow.
+   *
+   * @param {String} customerId The customerId to set for this device fingerprint. Optional if setCustomerId called in advance.
+   * @param {Function} callback Optional callback function to execute upon completion, passing an error if one occured.
+   * @example
+   * // if setCustomerId was already called
+   * ravelinjs.trackFingerprint();
+   * // else, you must inform Ravelin of the customerId explicitly
+   * ravelinjs.trackFingerprint('customer123');
+   */
+  RavelinJS.prototype.trackFingerprint = function(cb) {
```

Comment:
Ignore this section for now https://docs.google.com/document/d/18vwfbrQ7EzHHDjhzf_jeBVUVu4pA945VDBTvCaxSYTM/edit